src/v/cluster/archival_metadata_stm.cc

```diff
@@ -620,9 +620,145 @@ ss::future<std::error_code> archival_metadata_stm::process_anomalies(
+
+        if (!_raft->is_leader()) {
+            vlog(
+              _logger.info,
+              "Archival STM sync from offset {} to offest {} in "
+              "term {} was successful, but leadership moved",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term);
+            co_return false;
+        }
+
+        if (current_term != _raft->term()) {
+            vlog(
+              _logger.info,
+              "Archival STM sync from offset {} to offest {} in "
+              "term {} was successful, but term changed to {}",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term,
+              _raft->term());
+            co_return false;
+        }
+
+        if (
+          _last_replicated_offset
+          && _manifest->get_insync_offset() < _last_replicated_offset) {
+            vlog(
+              _logger.warn,
+              "Archival STM sync from offset {} to offest in {}"
+              "term {} was successful, but the last replicated offset moved to "
+              "{}",
+              last_applied_offset,
+              _last_replicated_offset,
+              current_term,
+              *_last_replicated_offset);
+            co_return false;
+        }
+
+        co_return true;
+    } else {
+        vlog(
+          _logger.debug,
+          "Syncing archival STM to latest term {}",
+          _raft->term());
+        const bool synced = co_await persisted_stm::sync(timeout);
+
+        if (!synced) {
+            vlog(_logger.debug, "Failed to sync archival STM to latest term");
+            co_return false;
+        }
+
+        if (!_raft->is_leader()) {
+            vlog(
+              _logger.debug,
+              "Archival STM sync to latest term was successful, but leadership "
+              "moved");
+            co_return false;
+        }
+
+        if (_insync_term != _raft->term()) {
+            vlog(
+              _logger.debug,
+              "Archival STM sync to latest term {} was successful, but the "
+              "term changed {}",
+              _insync_term,
+              _raft->term());
+            co_return false;
+        }
+
+        if (
+          _last_replicated_offset
+          && _manifest->get_insync_offset() < _last_replicated_offset) {
+            vlog(
+              _logger.debug,
+              "Archival STM sync to latest term {} was successful, but the "
+              "last applied offset {} is still below the last replicated "
+              "offset {}",
+              _insync_term,
+              _manifest->get_insync_offset(),
+              _last_replicated_offset);
+            co_return false;
+        }
+
+        co_return true;
+    }
+}
+
 ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
   model::record_batch batch, ss::abort_source& as) {
-    auto current_term = _insync_term;
+    const auto batch_last_offset = batch.last_offset();
```

Comment:
I thought the offset is assigned and serialized by Raft in the process of replicating. Perhaps we can get the resulting offset from the replicate() call?

Reply:
Yeah, that was a bug. Fixed now.